lib/ansible/vars/manager.py

```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
             self._vars_cache[host] = combine_vars(self._vars_cache[host], {varname: value})
         else:
             self._vars_cache[host][varname] = value
+
+
+class VarsWithSources(dict):
+    '''
+    Dict-like class for vars that also provides source information for each var
+    '''
+    def __init__(self, data, sources):
+        super(VarsWithSources, self).__init__(data)
+        self.sources = sources
+
+    def get_source(self, key):
+        return self.sources.get(key, None)
+
+    def __getitem__(self, key):
```

Comment:
This is actually one reason you want to use `MutableMapping` over subclassing `dict`.  `.get()` does not use `.__getitem__()` in `dict`. Both methods duplicate code for performance:

```pycon
>>> class VarsWithSources(dict):
...     def __getitem__(self, key):
...         val = super(VarsWithSources, self).__getitem__(key)
...         print(('__getitem__', val))
...         return val
...
>>> v = VarsWithSources()
>>> v['foo'] = 'bar'
>>> v['foo']
('__getitem__', 'bar')
'bar'
>>> v.get('foo')
'bar'
```

Notice that `.get()` doesn't cause the print to happen.  As such, your implementation as is will only work with some accesses of keys, and not others.